File: Java-CodeServer/src/main/resources/js/prompto/grammar/IdentifierList.js
var ObjectList = require("../utils/ObjectList").ObjectList;
var Dialect = require("../parser/Dialect").Dialect;

function IdentifierList(item) {
	ObjectList.call(this);
	item = item || null;
	if(item!==null) {
		this.add(item);
	}
	return this;
}

IdentifierList.prototype = Object.create(ObjectList.prototype);
IdentifierList.prototype.constructor = IdentifierList;

IdentifierList.parse = function(ids) {
	var parts = ids.split(",");
	var result = new IdentifierList();
	for(var i=0;i<parts.length;i++) {
		result.add(parts[i]);
	}
	return result;
};

IdentifierList.prototype.toDialect = function(writer, finalAnd) {
    finalAnd = finalAnd || false;
    switch(writer.dialect) {
        case Dialect.E:
            this.toEDialect(writer, finalAnd);
            break;
        case Dialect.O:
            this.toODialect(writer);
            break;
        case Dialect.S:
            this.toSDialect(writer);
            break;
    }
};

IdentifierList.prototype.toEDialect = function(writer, finalAnd) {
    switch(this.length) {
        case 0:
            return;
        case 1:
            writer.append(this[0]);
            break;
        default:
            for(var i=0;i<this.length;i++) {
                if(finalAnd && i==this.length-1)
                    break;
                writer.append(this[i]);
                writer.append(", ");
            }
            writer.trimLast(2);
            if(finalAnd) {
                writer.append(" and ");
                writer.append(this[this.length-1]);
            }
    }
};

IdentifierList.prototype.toODialect = function(writer) {
    if(this.length>0) {
        for(var i=0;i<this.length;i++) {
            writer.append(this[i]);
            writer.append(", ");
        }
        writer.trimLast(2);
    }
};

IdentifierList.prototype.toSDialect = function(writer) {
    this.toODialect(writer);
};

exports.IdentifierList = IdentifierList;

File: Java-Core/src/main/java/prompto/utils/IdentifierList.java
package prompto.utils;

import prompto.grammar.Identifier;


public class IdentifierList extends ObjectList<Identifier> {

	private static final long serialVersionUID = 1L;

	public static IdentifierList parse(String ids) {
		String[] parts = ids.split(",");
		IdentifierList result = new IdentifierList();
		for(String part : parts)
			result.add(new Identifier(part));
		return result;
	}
	
	public IdentifierList() {		
	}
	
	public IdentifierList(Identifier item) {
		this.add(item);
	}

	
	public void toDialect(CodeWriter writer, boolean finalAnd) {
		switch(writer.getDialect()) {
		case E:
			toEDialect(writer, finalAnd);
			break;
		case O:
			toODialect(writer);
			break;
		case S:
			toSDialect(writer);
			break;
		}
		
	}

	private void toEDialect(CodeWriter writer, boolean finalAnd) {
		switch(this.size()) {
		case 0:
			return;
		case 1:
			writer.append(this.getFirst());
			break;
		default:
			for(Identifier s : this) {
				if(finalAnd && s==this.getLast())
					break;
				writer.append(s);
				writer.append(", ");
			}
			writer.trimLast(2);
			if(finalAnd) {
				writer.append(" and ");
				writer.append(this.getLast());
			}
		}
	}

	private void toODialect(CodeWriter writer) {
		if(this.size()>0) {
			for(Identifier s : this) {
				writer.append(s);
				writer.append(", ");
			}
			writer.trimLast(2);
		}
	}

	private void toSDialect(CodeWriter writer) {
		toODialect(writer);
	}
}
